### PR TITLE
Implement MSC2290

### PR DESCRIPTION
**Pull request by anoadragon453**
Requires #6042 

Implements [MSC2290](https://github.com/matrix-org/matrix-doc/pull/2290). This PR adds two new endpoints, `/unstable/account/3pid/add` and `/unstable/account/3pid/bind`. Depending on the progress of that MSC the unstable prefix may go away.

This PR also removes the blacklist on some 3PID tests which occurs in #6042, as the corresponding Sytest PR changes them to use the new endpoints.

Finally, it also modifies the account deactivation code such that it doesn't just try to deactivate 3PIDs that were bound to the user's account, but any 3PIDs that were bound through the homeserver on that user's account.

SyTest PR: https://github.com/matrix-org/sytest/pull/703

The base branch is currently #6042, this will be changed to develop when #6042 merges.

### CODE REVIEW

**Comment by jryans** [comment]
Should this PR also edit the upgrade notes? The notes so far only mention registration and password reset being affected by email config, but now adding 3PIDs to the HS is also affected as well.

**Comment by anoadragon453** [comment]
I believe that's happening in #6033

**Comment by anoadragon453** [comment]
Something's happened with the commits in this branch and I suspect a force-push will be necessary.

File: synapse/rest/client/v2_alpha/account.py [hunk 1]
```diff
@@ -716,6 +698,74 @@ def on_POST(self, request):
         return 200, {}
 
 
+class ThreepidAddRestServlet(RestServlet):
+    PATTERNS = client_patterns("/account/3pid/add$", unstable=True)
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/account.py]:
```suggestion
    PATTERNS = client_patterns("/account/3pid/add$", releases=(), unstable=True)
```

File: synapse/rest/client/v2_alpha/account.py [hunk 2]
```diff
@@ -716,6 +698,74 @@ def on_POST(self, request):
...
+
+        address, _, medium, _, _, validated_at = validation_session
+
+        yield self.auth_handler.add_threepid(user_id, medium, address, validated_at)
+
+        return 200, {}
+
+
+class ThreepidBindRestServlet(RestServlet):
+    PATTERNS = client_patterns("/account/3pid/bind$", unstable=True)
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/account.py]:
```suggestion
    PATTERNS = client_patterns("/account/3pid/bind$", releases=(), unstable=True)
```

File: synapse/rest/client/v2_alpha/account.py [hunk 3]
```diff
@@ -690,6 +703,115 @@ def _add_threepid_to_account(self, user_id, validation_session):
...
+
+        assert_params_in_dict(body, ["client_secret", "sid"])
+        client_secret = body["client_secret"]
+        sid = body["sid"]
+        threepid_creds = {"client_secret": client_secret, "sid": sid}
+
+        requester = yield self.auth.get_user_by_req(request)
+        user_id = requester.user.to_string()
+
+        # We don't actually know which medium this 3PID is. Thus we first assume it's email,
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: hum, I'm not really happy with the duplication with /account/3pid, even if it is only temporary. Even temporary things have a habit of sticking around for long enough to be a pain. Can you factor out the duplication? The easiest way might be to move everything below this point out to the identity handler.

File: synapse/storage/registration.py [hunk 4]
```diff
@@ -586,6 +586,14 @@ def add_user_bound_threepid(self, user_id, medium, address, id_server):
             desc="add_user_bound_threepid",
         )
 
+    def user_get_bound_threepids(self, user_id):
+        return self._simple_select_list(
+            "user_threepid_id_server",
+            {"user_id": user_id},
```
Comment (richvdh) [inline synapse/storage/registration.py]: could you use keyword args here, for clarity?

File: synapse/handlers/deactivate_account.py [hunk 5]
```diff
@@ -73,7 +73,13 @@ def deactivate_account(self, user_id, erase_data, id_server=None):
         # unbinding
         identity_server_supports_unbinding = True
 
-        threepids = yield self.store.user_get_threepids(user_id)
+        # Retrieve the 3PIDs added to the user's account
+        user_threepids = yield self.store.user_get_threepids(user_id)
+
+        # Retrieve the 3PIDs this user has bound to an identity server
+        bound_threepids = yield self.store.user_get_bound_threepids(user_id)
+        threepids = user_threepids + bound_threepids
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/deactivate_account.py]: Ah, yes they certainly could contain the same threepids I think. Deduping is a good idea.

File: synapse/handlers/deactivate_account.py [hunk 6]
```diff
@@ -73,7 +73,13 @@ def deactivate_account(self, user_id, erase_data, id_server=None):
         # unbinding
         identity_server_supports_unbinding = True
 
-        threepids = yield self.store.user_get_threepids(user_id)
+        # Retrieve the 3PIDs added to the user's account
+        user_threepids = yield self.store.user_get_threepids(user_id)
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/deactivate_account.py]: Yeah, that probably doesn't make any sense in hindsight.

File: synapse/rest/client/v2_alpha/account.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: Yes, sorry. It was temporary in the sense that I was waiting for #6042 to land.

**Comment by anoadragon453** [comment]
Apologies for the force-push, needed to rebase onto develop.

File: synapse/rest/client/v2_alpha/account.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: Have been refactored out into a new method in IdentityHandler.

File: synapse/handlers/identity.py [hunk 7]
```diff
@@ -447,6 +418,50 @@ def requestMsisdnToken(
...
+
+        Returns:
+            Dict[str, str|int] if validation was successful, otherwise None
+        """
+        # We don't actually know which medium this 3PID is. Thus we first assume it's email,
+        # and if validation fails we try msisdn
+        validation_session = None
+
+        # XXX: We should need to keep wrapping and unwrapping this value
+        threepid_creds = {"client_secret": client_secret, "sid": sid}
```
Comment (anoadragon453) [inline synapse/handlers/identity.py]: I'll address this in a future PR

File: synapse/storage/registration.py [hunk 8]
```diff
@@ -586,6 +586,14 @@ def add_user_bound_threepid(self, user_id, medium, address, id_server):
             desc="add_user_bound_threepid",
         )
 
+    def user_get_bound_threepids(self, user_id):
```
Comment (richvdh) [inline synapse/storage/registration.py]: docstring please

File: synapse/handlers/identity.py [hunk 9]
```diff
@@ -447,6 +417,50 @@ def requestMsisdnToken(
...
+        # and if validation fails we try msisdn
+        validation_session = None
+
+        # XXX: We shouldn't need to keep wrapping and unwrapping this value
+        threepid_creds = {"client_secret": client_secret, "sid": sid}
+
+        # Try to validate as email
+        if self.hs.config.threepid_behaviour_email == ThreepidBehaviour.REMOTE:
+            # Ask our delegated email identity server
+            validation_session = yield self.threepid_from_creds(
```
Comment (richvdh) [inline synapse/handlers/identity.py]: we seem to have lost the HTTP error code handling? (should it be in `threepid_from_creds` anyway?)

Comment (anoadragon453) [reply synapse/handlers/identity.py]: Aw man, that's sad. Will re-add.

Comment (anoadragon453) [reply synapse/handlers/identity.py]: Ah no, it's here just removed in the diff. yay

**Comment by anoadragon453** [comment]
The following PR should help with the coverage a bit: https://github.com/matrix-org/synapse/pull/6083

But ultimately we should figure out why the code paths in IdentityHandler are apparently not being tested (even though I'm pretty sure our sytest MTA thing would test at least the email stuff)?

File: synapse/handlers/identity.py [hunk 10]
```diff
@@ -447,6 +425,64 @@ def requestMsisdnToken(
...
+                self.hs.config.account_threepid_delegate_email, threepid_creds
+            )
+        elif self.hs.config.threepid_behaviour_email == ThreepidBehaviour.LOCAL:
+            # Get a validated session matching these details
+            validation_session = yield self.store.get_threepid_validation_session(
+                "email", client_secret, sid=sid, validated=True
+            )
+
+        # Old versions of Sydent return a 200 http code even on a failed validation check.
```
Comment (richvdh) [inline synapse/handlers/identity.py]: this comment should be in threepid_from_creds now

File: synapse/handlers/identity.py [hunk 11]
```diff
@@ -447,6 +425,64 @@ def requestMsisdnToken(
...
+        # Thus, in addition to the HttpResponseException check above (which checks for
+        # non-200 errors), we need to make sure validation_session isn't actually an error,
+        # identified by containing an "error" key
+        # See https://github.com/matrix-org/sydent/issues/215 for details
+        if validation_session and "error" not in validation_session:
```
Comment (richvdh) [inline synapse/handlers/identity.py]: and afaict the check on `error` is redundant

File: synapse/storage/registration.py [hunk 12]
```diff
@@ -587,6 +587,18 @@ def add_user_bound_threepid(self, user_id, medium, address, id_server):
...
     def user_get_bound_threepids(self, user_id):
+        """Get the threepids that a user has bound to an identity server through the homeserver
+        The homeserver remembers where binds to an identity server occurred. Using this
+        method can retrieve those threepids.
+
+        Args:
+            user_id (str): The ID of the user to retrieve threepids for
+
+        Returns:
+            list[dict(str, str)]: List of dictionaries containing the following:
```
Comment (richvdh) [inline synapse/storage/registration.py]: Deferred

File: synapse/handlers/identity.py [hunk 13]
```diff
@@ -108,36 +79,52 @@ def threepid_from_creds(self, id_server, creds):
...
+            logger.info(
+                "%s returned %i for threepid validation for: %s",
+                id_server,
+                e.code,
+                creds,
+            )
+            return None
+
+        # Old versions of Sydent return a 200 http code even on a failed validation
+        # check. # Thus, in addition to the HttpResponseException check above (which
```
Comment (jryans) [inline synapse/handlers/identity.py]:
```suggestion
        # check. Thus, in addition to the HttpResponseException check above (which
```

Comment (richvdh) [reply synapse/handlers/identity.py]: thanks :)

**Pull request merged**